test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java

```diff
@@ -0,0 +1,11205 @@
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.MUL_V, IRNode.STORE_VECTOR},
+        applyIf = {"MaxVectorSize", "< 8"},
+        applyIfCPUFeature = {"asimd", "true"})
+    public static void testIntP0(int[] data) {
+        for (int j = 0; j < RANGE; j++) {
+            data[j + 0] = (int)(data[j] * (int)-11);
+        }
+    }
+
+    @Run(test = "testIntP0")
+    @Warmup(0)
+    public static void runIntP0() {
+        int[] data = new int[RANGE];
+        init(data);
+        testIntP0(data);
+        verify("testIntP0", data, goldIntP0);
+    }
+
+    @Test
+    // cpu: sse4.1 to avx -> vector_width: 16 -> elements in vector: 4
+    @IR(counts = {IRNode.LOAD_VECTOR, "> 0", IRNode.MUL_V, "> 0", IRNode.STORE_VECTOR, "> 0"},
+        applyIf = {"MaxVectorSize", ">= 8"},
+        applyIfCPUFeatureAnd = {"sse4.1", "true", "avx2", "false"})
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.MUL_V, IRNode.STORE_VECTOR},
+        applyIf = {"MaxVectorSize", "< 8"},
+        applyIfCPUFeatureAnd = {"sse4.1", "true", "avx2", "false"})
+    // cpu: avx2 -> vector_width: 32 -> elements in vector: 8
+    @IR(counts = {IRNode.LOAD_VECTOR, "> 0", IRNode.MUL_V, "> 0", IRNode.STORE_VECTOR, "> 0"},
+        applyIf = {"MaxVectorSize", ">= 8"},
+        applyIfCPUFeatureAnd = {"avx2", "true", "avx512", "false"})
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.MUL_V, IRNode.STORE_VECTOR},
+        applyIf = {"MaxVectorSize", "< 8"},
+        applyIfCPUFeatureAnd = {"avx2", "true", "avx512", "false"})
+    // cpu: avx512 -> vector_width: 64 -> elements in vector: 16
+    @IR(counts = {IRNode.LOAD_VECTOR, "> 0", IRNode.MUL_V, "> 0", IRNode.STORE_VECTOR, "> 0"},
+        applyIf = {"MaxVectorSize", ">= 8"},
+        applyIfCPUFeature = {"avx512", "true"})
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.MUL_V, IRNode.STORE_VECTOR},
+        applyIf = {"MaxVectorSize", "< 8"},
+        applyIfCPUFeature = {"avx512", "true"})
+    // cpu: asimd -> vector_width: 32 -> elements in vector: 8
+    @IR(counts = {IRNode.LOAD_VECTOR, "> 0", IRNode.MUL_V, "> 0", IRNode.STORE_VECTOR, "> 0"},
+        applyIf = {"MaxVectorSize", ">= 8"},
+        applyIfCPUFeature = {"asimd", "true"})
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.MUL_V, IRNode.STORE_VECTOR},
+        applyIf = {"MaxVectorSize", "< 8"},
+        applyIfCPUFeature = {"asimd", "true"})
+    public static void testIntM1(int[] data) {
+        for (int j = 1; j < RANGE; j++) {
+            data[j + -1] = (int)(data[j] * (int)-11);
+        }
+    }
+
+    @Run(test = "testIntM1")
+    @Warmup(0)
+    public static void runIntM1() {
+        int[] data = new int[RANGE];
+        init(data);
+        testIntM1(data);
+        verify("testIntM1", data, goldIntM1);
+    }
+
+    @Test
+    // cpu: sse4.1 to avx -> vector_width: 16 -> elements in vector: 4
+    // positive byte_offset 4 can lead to cyclic dependency
+    @IR(counts = {IRNode.LOAD_VECTOR, "> 0", IRNode.MUL_V, "> 0", IRNode.STORE_VECTOR, "> 0"},
```

Comment:
Needs to fix the test generation script. This IR rule looks incorrect since there is a valid dependency with distance 1.

Reply:
@jatin-bhateja I think the IR rule is just ineffective. I have the following condition in it that will never be met:
`applyIfAnd = {"MaxVectorSize", ">= 8", "MaxVectorSize", "<= 4"},`
The `<= 4` must hold so that `byte_offset <= MaxVectorSize`. But `>= 8` must hold so that two ints fit in a vector.

I could improve the script and filter out such ineffective IR rules. Not sure if that is worth it though.

Reply:
With +AlignVector behavior with and without Vectorize,true pragma should match.

```
static void test1() {
     for (int i = 4; i < 100; i++) { 
         fArr[i + 4] = fArr[i];
     }
} 
```
  
```
CPROMPT>javad -XX:+TraceNewVectors -XX:+AlignVector -XX:-TraceSuperWord -cp . bug
WARNING: Using incubator modules: jdk.incubator.vector
res = 0.0
CPROMPT>
CPROMPT>javad -XX:+TraceNewVectors -XX:+AlignVector -XX:-TraceSuperWord -XX:CompileCommand=Vectorize,bug::test1,true -cp . bug
CompileCommand: Vectorize bug.test1 bool Vectorize = true
WARNING: Using incubator modules: jdk.incubator.vector
new Vector node:  990  LoadVector  === 373 856 824  [[ 822 802 800 798 718 706 556 196 ]]  @float[int:>=0]:NotNull:exact+any *, idx=6; mismatched #vectory[8]:{float} !orig=[823],[719],[557],[199],143 !jvms: bug::test1 @ bci:18 (line 7)
new Vector node:  991  StoreVector  === 855 856 825 990  [[ 988 195 856 ]]  @float[int:>=0]:NotNull:exact+any *, idx=6; mismatched  Memory: @float[int:>=0]:NotNull:exact+any *, idx=6; !orig=[822],[718],[556],[196],164 !jvms: bug::test1 @ bci:19 (line 7)
res = 0.0
```

Reply:
@jatin-bhateja 
Under `aarch64`, I have made bad experiences with `SuperWordMaxVectorSize`. It is not properly adjusted to be at most `MaxVectorSize`. For example if the `aarch64` machine only supports `MaxVectorSize <= 32`, but I set `SuperWordMaxVectorSize = 64`, then it will keep it at `64`. So then my IR rules fail.
For the `x86 / x64` machines we have:
https://github.com/openjdk/jdk/blob/33bec207103acd520eb99afb093cfafa44aecfda/src/hotspot/cpu/x86/vm_version_x86.cpp#L1314-L1333

@fg1417 Would you like to implement this for `aarch64`?

Reply:
@jatin-bhateja 
> Thanks, even though newly added test now passes at all AVX and SSE level can you kindly investigate why should following be vectorized with un-aligned accesses when it carries a cross iteration true dependency with distance 4.

The cyclic dependency is at a distance of 3, not 4 in this example. Ints are 4 bytes. Thus, the `byte_offset` is 12 bytes. So if `MaxVectorSize <= 12`, we cannot ever have a cyclic dependency within a vector.

See my explanations at the beginning of the test file, for example:
https://github.com/openjdk/jdk/blob/fb7f6dd9fbc2a6086d2ad36e0681fbc9eff6c9a7/test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java#L49-L57

Reply:
You can see that the rules for distance 4 are adjusted, if you look at `testIntP4`:
https://github.com/openjdk/jdk/blob/fb7f6dd9fbc2a6086d2ad36e0681fbc9eff6c9a7/test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java#L1202-L1230

Reply:
Correct, my bad just pasted a wrong example!,  Do you see any value in 

> @jatin-bhateja
> 
> > Thanks, even though newly added test now passes at all AVX and SSE level can you kindly investigate why should following be vectorized with un-aligned accesses when it carries a cross iteration true dependency with distance 4.
> 
> The cyclic dependency is at a distance of 3, not 4 in this example. Ints are 4 bytes. Thus, the `byte_offset` is 12 bytes. So if `MaxVectorSize <= 12`, we cannot ever have a cyclic dependency within a vector.
> 
> See my explanations at the beginning of the test file, for example:
> 
> https://github.com/openjdk/jdk/blob/fb7f6dd9fbc2a6086d2ad36e0681fbc9eff6c9a7/test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java#L49-L57

My bad, just pasted a wrong example, do you really see any value in generating tests for synthetic vector sizes which are non-power of two. Lets remove them to reduce noise ?

Reply:
> do you really see any value in generating tests for synthetic vector sizes where MaxVectorSize is a non-power of two. Lets remove them to reduce noise ?

I see a value in having non-power of 2 offsets, yes. They should vectorize if the vector width is small enough. And then there are some values like `18, 20, 192` that are a there to check vectorization with `+AlignVector`.

Maybe you find the `MaxVectorSize <= 12` "noisy" somehow, because it is equivalent to `MaxVectorSize <= 8`? I find it rather helpful, because `12` reflects the `byte_offset`, and so makes the rule a bit more understandable.

Reply:
> > With +AlignVector behavior with and without Vectorize,true pragma should match.
> 
> This was about example with `fArr[i + 4] = fArr[i];` in the loop. `byte_offset = 4 * 4 = 16`.
> 
> @jatin-bhateja I am not sure what you are trying to say, what do you mean by `should match`?
> 

Yes, this was a bug in mainline where we were incorrectly vectorizing which is now fixed with your changes, just wanted to get that point highlighted.

Reply:
> @jatin-bhateja Under `aarch64`, I have made bad experiences with `SuperWordMaxVectorSize`. It is not properly adjusted to be at most `MaxVectorSize`. For example if the `aarch64` machine only supports `MaxVectorSize <= 32`, but I set `SuperWordMaxVectorSize = 64`, then it will keep it at `64`. So then my IR rules fail. For the `x86 / x64` machines we have:
> 
> https://github.com/openjdk/jdk/blob/33bec207103acd520eb99afb093cfafa44aecfda/src/hotspot/cpu/x86/vm_version_x86.cpp#L1314-L1333
> 
> @fg1417 Would you like to implement this for `aarch64`?

@eme64 thanks for pointing it out! I'll take a look at it.